Version 2 upgrade of go-crypto

lubux opened the pull request:
This pull request includes the version 2 upgrade in a subfolder in openpgp and the v6 changes from #161.
See [here](https://github.com/ProtonMail/go-crypto/compare/update-to-v6...lubux:go-crypto:version-2) for the v2 changes history.

# Version 2 Update

This document summarizes the major changes for reading and writing OpenPGP message introduced by `ProtonMail/go-crypto/openpgp/v2` compared to `ProtonMail/go-crypto/openpgp` with the v6 update (#161.).


## Import

The v2 API can be imported as:
```
openpgp ProtonMail/go-crypto/openpgp/v2
```

## API breaking changes

`openpgp.v2.Entity` struct fields have changed:
- `SelfSignature *packet.Signature` removed
- `Signatures []*packet.Signature` removed
- `DirectSignatures []*packet.VerifiableSignature` added
- `Revocations []*packet.VerifiableSignature` changed type
  
`openpgp.v2.Entity` changed API methods:
- `PrimaryIdentity(date time.Time)` has a time argument now.
- `EncryptionKey(date time.Time, config *packet.Config)` has a new config argument.
- `CertificationKey(date time.Time, config *packet.Config)` has a new config argument.
- `CertificationKeyById(date time.Time, id uint64, config *packet.Config)` has a new config argument.
- `SigningKey(date time.Time, config *packet.Config)` has a new config argument.
- `SigningKeyById(date time.Time, id uint64, config *packet.Config)` has a new config argument.
- `Revoke(reason packet.ReasonForRevocation, reasonText string, config *packet.Config)` changed name instead of RevokeKey.

`openpgp.v2.Entity` removed API methods:
- `RevokeSubkey(...) ` replaced by `(Subkey).Revoke(...)`

`openpgp.v2.Subkey` struct fields have changed:
- `Sig *packet.Signature` removed
- `Bindings []*packet.VerifiableSignature` added
- `Primary *Entity` added, points to the primary key.
- `Revocations []*packet.VerifiableSignature` changed type

`openpgp.v2.Subkey` changed API methods:
- `Revoked(selfCertification *packet.Signature, date time.Time)` has a new selfCertification argument, which points to the self signature to be used.

`openpgp.v2.Identity` struct fields have changed:
- `SelfSignature *packet.Signature` removed
- `Signatures []*packet.Signature ` removed
- `SelfCertifications []*packet.VerifiableSignature` added
- `OtherCertifications []*packet.VerifiableSignature` added
- `Primary *Entity` added, points to the primary key.
- `Revocations []*packet.VerifiableSignature` changed type

`openpgp.v2.Identity` changed API methods:
- `Revoked(selfCertification *packet.Signature, date time.Time)` has a new selfCertification argument, which points to the self signature to be used.

`openpgp.v2.Key` struct fields have changed:
- `PrimarySelfSignature *packet.Signature ` added, which points to the selected self signature of the primary key.
- `Revocations []*packet.VerifiableSignature` changed type

`openpgp.v2.KeyRing` interface has has changed:
- `KeysByIdUsage(...)` removed
- `DecryptionKeys(...)` removed
- `EntitiesById(id uint64) []*Entity` added. This is the main internal method to access keys from the keyring now.
  
`openpgp.v2.FileHints` struct field has changed:
- `IsBinary` removed  and `IsUTF8` added 
  
`openpgp.v2` API changes for reading messages:
- `VerifyDetachedSignatureAndHash(...)` removed, headers in clearsigned messages are no longer checked.
- `VerifyDetachedSignatureAndSaltedHash(...)` removed
- `CheckDetachedSignature(...)` removed, call `VerifyDetachedSignature(...)` instead
- `CheckDetachedSignatureAndSaltedHash(...)` removed 
- `CheckDetachedSignatureAndHash(...)` removed
- `CheckArmoredDetachedSignature` removed call `VerifyArmoredDetachedSignature` instead

`openpgp.v2` API changes for writing messages:
- `DetachSign(..., signers []*Entity,...)` takes now a slice of entities instead of a single entity as an argument.
- `ArmoredDetachSign(..., signers []*Entity,..., , params *SignParams)` takes now a slice of entities instead of a single entity as an argument and replaces arguments with a SignParams object.
- `DetachSignText(..., signers []*Entity,...)` takes now a slice of entities instead of a single entity as an argument.
- `ArmoredDetachSignText(..., signers []*Entity,...)` takes now a slice of entities instead of a single entity as an argument.
- `EncryptText(...)` removed call `EncryptWithParams(...)` instead
- `EncryptSplit(...)` removed call `EncryptWithParams(...)` instead
- `EncryptTextSplit(...)` removed call `EncryptWithParams(...)` instead
- `Encrypt(..., toHidden []*Entity, signers []*Entity)` now takes an additional toHidden recipients argument and takes now a slice of signer entities instead of a single entity as an argument.
- `Sign(..., signers []*Entity,...)` takes now a slice of entities instead of a single entity as an argument.

## Features added

### Intended recipients

Version 2 of the ProtonMail/go-crypto library introduces a feature for including the recipients' key fingerprints in signatures during message encryption.
When encrypting and signing a message, the intended recipients are automatically included in the signature, unless specifically hidden (i.e., hidden recipients). 
During the decryption process, if the signature contains intended recipients and the appropriate configuration flag is set, the library verifies whether the primary ID of the decryption key is present in the recipient list. 
This check can be disabled in the config when a hidden recipient decrypts the message.

### Multi-signature support

In previous iterations of ProtonMail/go-crypto, only a single signature creation and verification were supported in a PGP message. 
However, in Version 2, the library introduces the ability to sign messages with multiple signatures using different keys, such as a v4 and a v6 key.
The encryption and signing methods now accept multiple signing keys as arguments, with each key designated for a specific signature. 
When reading PGP messages with Version 2, the library maintains an internal state for each known signature and verifies all of them within the message.
To facilitate this functionality, the message details struct includes a new field that stores the verification state for each signature. A message is considered valid if at least one of the signatures successfully validates without any errors.
For callers, the process of checking for signature errors remains similar to previous versions. 
However, if the caller requires the verification state of all signatures, they can utilize the new field in the message details struct.

### Rework how signatures in keys and signatures are verified

In previous iterations of ProtonMail/go-crypto, key verification occurred during import based on the current time, while signature verification did not involve further key checks. 
However, this approach had limitations, as invalid keys could have been valid at the time of signature creation and mistakenly considered invalid.

Version 2 changes how and when signatures are verified in keys (i.e., direct-signatures, self-signatures of userids, binding signatures in subkeys, revocations, etc).
Unlike before, key signature verification no longer takes place during parsing. 
Instead, keys are now validated when they are utilized, following a similar approach to key handling in OpenPGP.js. 
Additionally, all signatures and expirations are validated to adhere to the key lifecycle outlined in the RFC.
The validity of keys can now be checked at different points in time, leading to the following specific modifications:
- During entity parsing, key validity is not checked.
- When used for encryption or signing, keys are verified using the current time during the writing process.
- During reading, the library verifies that each verification key was valid at the time of signature creation.
- A clear separation is maintained between Entity, Subkey, Identity, and their respective validation methods.
- Signature verification results are cached and reused to optimize computation.

Further, version 2 includes various small improvements to increase the robustness of the key parsing functions. 

### Weak algorithm rejection

Version 2 introduces the option to specify weak algorithms for signatures in the config.
Signatures that use weak algorithms are considered invalid.

### Optional packet sequence checker

Version 2 introduces a new feature that enables the validation of packet sequences in PGP messages. 
This functionality can be enabled in the config struct.
In particular, it implements the pushdown automata (PDA) from PGPainless, developed by Paul Schaub. 
By leveraging this feature, users can ensure that the packet sequences in their PGP messages are valid and comply with the required structure. 
This addition further enhances the overall reliability and security of PGP message handling in Version 2.

### Session key encryption and decryption

Version 2 allows advanced users to retrieve the session key while encrypting a message by setting the respective flag in the config.
In decryption, a caller can provide a session key that should be used for decryption.

### Unify write/read API

Version 2 improves the compatibility between different APIs to allow combinations. 
The `DetachSign` function requires the caller to provide a `Reader` for the message, while
encrypt returns a `WriteCloser` to which the message is written to.
The new version adds a function `DetachSignWriter`, which returns a `WriteCloser` similar to the encryption API.
On the reading side, the verify detached signature API now relies on the
same signature verification logic as the other read functions. 
Additionally, a new `VerifyDetachedSignatureReader` method similar to the `ReadMessage` API is introduced.
It returns a message details struct that once read verifies the signature.
Allows to chain different readers from the API, for example, to have a streaming API for encrypted detached signatures.

### Params struct as a function argument in the write API 

With the inclusion of new features, the write functions in go-crypto experienced significant growth in numbers. Each combination has its dedicated function. 
Version 2 introduces an `EncryptWithParams`/`SignWithParams` function that takes an `EncryptParams`/`SignParams` struct as an argument. The struct allows configuring the different features. 
This approach effectively reduces the number of API methods and simplifies the process of adding new features while maintaining compatibility with previous versions.

### Others

- Disable armor checksum on default `armor.Encode` method
- Make `unarmor` more robust to empty header values 
- Allow key generation of v6 keys without a `Identity`
- Allow compression in inline signed messages
- Consider key preferences in detached signatures
- Only compare time at a second granularity 
- Signal if tag is not verified on close in aead decryption
- Ensure that critical unknown packet tags result in message rejection
- Ensure that decompression streams are closed and that the packet is completely read
- Ensure that entity parsing does not reject keys with unknown subkeys 
- Check for known curves early when parsing ECDSA and ECDH keys
- Skip signatures with the wrong type while parsing an entity
- Support for signatures that appear in front of the data
- Change file hints field IsBinary to IsUTF8

wussler commented on openpgp/v2/write.go:
```diff
@@ -0,0 +1,1033 @@
+	IsUTF8 bool
+	// FileName hints at the name of the file that should be written. It's
+	// truncated to 255 bytes if longer. It may be empty to suggest that the
+	// file should not be written to disk. It may be equal to "_CONSOLE" to
```
This comment is outdated

lubux replied:
3db51732b5b372aebda2697e4cdccbbbf3bb2fec

wussler commented on openpgp/v2/write.go:
```diff
@@ -0,0 +1,1033 @@
+			if h, ok := algorithm.HashIdToHash(acceptableHashes[0]); ok {
+				hash = h
+			} else {
+				return 0, errors.InvalidArgumentError("no candidate hash functions are compiled in.")
+			}
+		} else {
+			return 0, errors.InvalidArgumentError("no candidate hash functions are compiled in.")
```
```suggestion
				return 0, errors.UnsupportedError("no candidate hash functions available")
			}
		} else {
			return 0, errors.UnsupportedError("no candidate hash functions available")
```

lubux replied:
3db51732b5b372aebda2697e4cdccbbbf3bb2fec

wussler commented on openpgp/v2/write.go:
```diff
@@ -0,0 +1,1033 @@
+			} else if curve == packet.CurveBrainpoolP384 ||
+				curve == packet.CurveNistP384 {
+				return []uint8{
+					hashToHashId(crypto.SHA256),
```
Why SHA256 but not SHA3-256?

lubux replied:
Good catch, this is a mistake. SHA256 should not be allowed with 384-bit curves.

Fixed in be8d42d6408deb29cdf255bcc71fa0bcbfdb00d5

wussler commented on openpgp/v2/write_test.go:
```diff
@@ -0,0 +1,994 @@
+}
+
+func TestNewEntity(t *testing.T) {
+
```
```suggestion
```

wussler commented on openpgp/v2/write_test.go:
```diff
@@ -0,0 +1,994 @@
+	}
+	_, ok := decPackets.(*packet.Compressed)
+	if !ok {
+		return errors.InvalidArgumentError("No compressed packets found")
```
StructuralError?

lubux replied:
Old code that I have not touched, but I can change that in both versions. 

1ab7fbc54e3e3a1d7504945b1f6ef005fb5472da

wussler commented on openpgp/write.go:
```diff
@@ -380,15 +398,19 @@ func encrypt(keyWriter io.Writer, dataWriter io.Writer, to []*Entity, signed *En
-		if !sig.SEIPDv2 {
+		primarySelfSignature, _ := to[i].PrimarySelfSignature()
+		if primarySelfSignature == nil {
+			return nil, errors.InvalidArgumentError("entity without a self-signature")
```
```suggestion
			return nil, errors.StructuralError("entity without a self-signature")
```

lubux replied:
1f9ab129a56b84efab282d651baf77c2dc4e72bb

wussler commented on openpgp/write.go:
```diff
@@ -465,7 +487,11 @@ func Sign(output io.Writer, signed *Entity, hints *FileHints, config *packet.Con
-	preferredHashes := signed.PrimaryIdentity().SelfSignature.PreferredHash
+	primarySelfSignature, _ := signed.PrimarySelfSignature()
+	if primarySelfSignature == nil {
+		return nil, errors.InvalidArgumentError("signed entity has no self-signature")
```
```suggestion
		return nil, errors.StructuralError("signed entity has no self-signature")
```

lubux replied:
1f9ab129a56b84efab282d651baf77c2dc4e72bb

wussler commented on openpgp/write_test.go:
```diff
@@ -498,6 +503,11 @@ func TestEncryption(t *testing.T) {
 		}
 
 		w, err := Encrypt(buf, kring[:1], signed, nil /* no hints */, config)
+		if err != nil && config.AEAD() != nil && !test.okV6 {
```
I think here we should assert whether `(err != nil) == (test.okV6 && config.AEAD())`

lubux replied:
59054bcfe7b22dd6249705e188335ccfea5d6ac9

wussler commented on openpgp/x25519/x25519.go:
```diff
@@ -0,0 +1,217 @@
+	return nil
+}
+
+// GenerateKey generates a new x25519 key pair
```
```suggestion
// GenerateKey generates a new x25519 key pair.
```

Top-level comments must end with a dot

lubux replied:
3198ed3fada927f01bd289c9e3339891f395df17

wussler commented on openpgp/x448/x448.go:
```diff
@@ -0,0 +1,217 @@
+	return nil
+}
+
+// GenerateKey generates a new x448 key pair
```
Same here

lubux replied:
49087ba51c25d500e1a8dfe519fc219979b1d416

wussler commented on openpgp/x448/x448.go:
```diff
@@ -0,0 +1,217 @@
+}
+
+// Encrypt encrpyts a sessionKey with x448 according to
+// the OpenPGP crypto refresh specification section 5.1.6. The function assumes that the
```
```suggestion
// the OpenPGP crypto refresh specification section 5.1.7. The function assumes that the
```

lubux replied:
49087ba51c25d500e1a8dfe519fc219979b1d416

wussler commented on openpgp/x448/x448.go:
```diff
@@ -0,0 +1,217 @@
+// sessionKey has the correct format and padding according to the specification.
+func Encrypt(rand io.Reader, publicKey *PublicKey, sessionKey []byte) (ephemeralPublicKey *PublicKey, encryptedSessionKey []byte, err error) {
+	var ephemeralPrivate, ephemeralPublic, staticPublic, shared x448lib.Key
+	// Check that the input static public key has 32 bytes
```
```suggestion
	// Check that the input static public key has 57 bytes
```

lubux replied:
49087ba51c25d500e1a8dfe519fc219979b1d416

wussler commented on openpgp/x448/x448.go:
```diff
@@ -0,0 +1,217 @@
+// private key and ephemeral public key
+func Decrypt(privateKey *PrivateKey, ephemeralPublicKey *PublicKey, ciphertext []byte) (encodedSessionKey []byte, err error) {
+	var ephemeralPublic, staticPrivate, shared x448lib.Key
+	// Check that the input ephemeral public key has 32 bytes
```
```suggestion
	// Check that the input ephemeral public key has 57 bytes
```

lubux replied:
0f7936b18645e8bf79b9d7134675580733d0490b

wussler replied:
Good job with 56 instead of 57. I mixed up Ed and X448.

lubux commented on openpgp/clearsign/clearsign.go:
```diff
@@ -199,10 +203,12 @@ func Decode(data []byte) (b *Block, rest []byte) {
+	hashers     []hash.Hash // one per key in privateKeys
+	hashType    crypto.Hash
+	toHash      io.Writer         // writes to all the hashes in hashers
+	salts       [][]byte          // salts for the signatures if v6
```
The headers are dropped in v6 but the v6 signatures still require a salt to be prepended and in the signature itself.

lubux commented on openpgp/clearsign/clearsign_test.go:
```diff
@@ -140,17 +182,17 @@ func (qr *quickRand) Read(p []byte) (int, error) {
+	for nKeys := 1; nKeys < 4; nKeys++ {
 	nextTest:
-		for nExtra := 0; nExtra < 4; nExtra++ {
+		for nExtra := 2; nExtra < 4; nExtra++ {
```
Good point, would make it nicer. I agree that we should do refactor of the tests in a separate PR, I did consider a refactor in tests when implementing v6.

lubux commented on openpgp/armor/armor.go:
```diff
@@ -142,19 +104,12 @@ func (l *lineReader) Read(p []byte) (n int, err error) {
 // a running CRC of the resulting data and checks the CRC against the value
```
Addressed in 4aabb698d496590e16f6219a55f82bf420f38003

lubux commented on openpgp/armor/encode.go:
```diff
@@ -14,6 +14,24 @@ var blockEnd = []byte("\n=")
 
+const crc24Init = 0xb704ce
+const crc24Poly = 0x1864cfb
+const crc24Mask = 0xffffff
```
Addressed in https://github.com/ProtonMail/go-crypto/commit/4aabb698d496590e16f6219a55f82bf420f38003

lubux commented on openpgp/ed25519/ed25519.go:
```diff
@@ -0,0 +1,99 @@
+)
+
+const PointSize = 32
+const PrivateKeySize = 64
```
Addressed with 820b3958eed84edd6cde073170a018d4a38c838c

lubux commented on openpgp/ed25519/ed25519_test.go:
```diff
@@ -0,0 +1,63 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(priv.Key) != 64 && len(priv.Point) != 32 {
```
Addressed with 820b3958eed84edd6cde073170a018d4a38c838c

lubux commented on openpgp/ed448/ed448_test.go:
```diff
@@ -0,0 +1,63 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(priv.Key) != 114 && len(priv.Point) != 57 {
```
13e46127d1eb4616696171ac6e3cac41e2cb0956

lubux commented on openpgp/ed448/ed448.go:
```diff
@@ -0,0 +1,103 @@
+	return pk.Key[:PointSize]
+}
+
+// MarshalByteSecret returns the underlying 32 byte seed of the private key
```
13e46127d1eb4616696171ac6e3cac41e2cb0956

lubux commented on openpgp/integration_tests/v2/utils_test.go:
```diff
@@ -0,0 +1,306 @@
+	return outString, nil
+}
+
+var runes = []rune("abcdefghijklmnopqrstuvwxyz0123456789ABCDEFGHIJKMNOPQRSTUVWXYZ.@-_:;?/!#$%^&*{}[]'\"+~()<>")
```
It seems not work.  Note that this file is just a copy of the old code for the new v2 api. I have not touched it. Might be good to do a refactor as well in a separate pull request.

wussler replied:
Then good like it is :+1:

lubux commented on openpgp/integration_tests/utils_test.go:
```diff
@@ -226,14 +228,16 @@ func randConfig() *packet.Config {
 		packet.CompressionNone,
-		packet.CompressionZIP,
-		packet.CompressionZLIB,
+		//packet.CompressionZIP,
```
Mistake fixed in 3d55a7e6fed727e54704208f2b2fb44cb497acb9

lubux commented on openpgp/keys_v5_test.go:
```diff
@@ -7,7 +7,6 @@ import (
 	"testing"
 
 	"github.com/ProtonMail/go-crypto/openpgp/armor"
```
Only the high-level api has a v2 the rest is shared and backwards compatible. i.e., there is no v2 armor package.

lubux commented on openpgp/packet/config.go:
```diff
@@ -110,6 +134,21 @@ type Config struct {
+	// a decryption key for an encrypted and signed messages should be checked
+	// to be present in the signatures intended recipient list.
+	// In encrypt and sign intended recipients are only included if this flag is true.
+	// if config is nil or flag is nil, it defaults to true
```
Improved :grin: 140825517cf4fe20d65246c9c12cfa55908f03e4

lubux commented on openpgp/packet/config.go:
```diff
@@ -54,8 +69,11 @@ type Config struct {
 	S2KCount int
-	// RSABits is the number of bits in new RSA keys made with NewEntity.
-	// If zero, then 2048 bit keys are created.
+	// An S2K specifier can be stored in the secret keyring to specify
```
Nope xD 140825517cf4fe20d65246c9c12cfa55908f03e4

lubux commented on openpgp/packet/encrypted_key.go:
```diff
@@ -114,51 +183,108 @@ func (e *EncryptedKey) Decrypt(priv *PrivateKey, config *Config) error {
+	if e.Version == 6 {
+		packetLen = 1 /* version */ + 1 /* algo */ + encodedLength + 1 /* key version */
+		if e.KeyVersion == 6 {
+			packetLen += 32
```
c10c41e5cf69280dc78b82a8e6689fc68f66957a

lubux commented on openpgp/packet/encrypted_key.go:
```diff
@@ -176,49 +302,143 @@ func (e *EncryptedKey) Serialize(w io.Writer) error {
+	return SerializeEncryptedKeyAEAD(w, pub, cipherFunc, config.AEAD() != nil, key, config)
+}
+
+// SerializeEncryptedKey serializes an encrypted key packet to w that contains
```
c10c41e5cf69280dc78b82a8e6689fc68f66957a

lubux commented on openpgp/packet/marker.go:
```diff
@@ -0,0 +1,33 @@
+
+const markerString = "PGP"
+
+// parse just ignores the padding content.
```
4863db4ca50f5dfe716f7ad8c53ab3609f3cf8f9

lubux commented on openpgp/packet/public_key.go:
```diff
@@ -600,7 +747,8 @@ func (pk *PublicKey) VerifySignature(signed hash.Hash, sig *Signature) (err erro
 	hashBytes := signed.Sum(nil)
-	if sig.Version == 5 && (hashBytes[0] != sig.HashTag[0] || hashBytes[1] != sig.HashTag[1]) {
+	// see discussion https://github.com/ProtonMail/go-crypto/issues/107
+	if sig.Version >= 5 && hashBytes[0] != sig.HashTag[0] || hashBytes[1] != sig.HashTag[1] {
```
Good catch, fixed in ddb84e3f58e191895bc0c0e56d7e9d8b898239dd

lubux commented on openpgp/packet/public_key.go:
```diff
@@ -786,21 +982,49 @@ func (pk *PublicKey) BitLength() (bitLength uint16, err error) {
 // expired or is created in the future.
 func (pk *PublicKey) KeyExpired(sig *Signature, currentTime time.Time) bool {
-	if pk.CreationTime.After(currentTime) {
+	if pk.CreationTime.Unix() > currentTime.Unix() {
```
We had issues, where time included nanoseconds for certain operations and comparisons failed where they should not have failed. 
I thought it would be cleaner and less error prone to only rely on unix timestamps in time comparisons since the OpenPGP standard also only uses unix timestamp granularity. 

(e.g., https://github.com/ProtonMail/gopenpgp/pull/240)

lubux commented on openpgp/packet/public_key.go:
```diff
@@ -447,10 +570,10 @@ func (pk *PublicKey) SerializeForHash(w io.Writer) error {
 		pLength += 10 // version, timestamp (4), algorithm, key octet count (4).
 		w.Write([]byte{
-			0x9A,
+			0x95 + byte(pk.Version),
```
1dc0808cd73465db36c96854ffcef2fa544e1686

lubux commented on openpgp/packet/public_key.go:
```diff
@@ -727,11 +897,8 @@ func (pk *PublicKey) VerifySubkeyRevocationSignature(sig *Signature, signed *Pub
-		return nil, errors.UnsupportedError("hash function")
-	}
-	h = hashFunc.New()
+func userIdSignatureHash(id string, pk *PublicKey, hash hash.Hash) (h hash.Hash, err error) {
```
Yes this was just lacy refactoring on my part.
Fixed in 1dc0808cd73465db36c96854ffcef2fa544e1686

lubux commented on openpgp/packet/public_key.go:
```diff
@@ -749,10 +916,39 @@ func userIdSignatureHash(id string, pk *PublicKey, hashFunc crypto.Hash) (h hash
 }
 
+// directSignatureHash returns a Hash of the message that needs to be signed
+func directKeySignatureHash(pk *PublicKey, hash hash.Hash) (h hash.Hash, err error) {
```
1dc0808cd73465db36c96854ffcef2fa544e1686

lubux commented on openpgp/packet/encrypted_key.go:
```diff
@@ -114,51 +183,108 @@ func (e *EncryptedKey) Decrypt(priv *PrivateKey, config *Config) error {
+				return errors.StructuralError("v3 PKESK mandates AES as cipher function for x25519 and x448")
+			}
+		}
+		key = b[:]
```
c10c41e5cf69280dc78b82a8e6689fc68f66957a

lubux commented on openpgp/packet/signature.go:
```diff
@@ -166,11 +214,21 @@ func (sig *Signature) parse(r io.Reader) (err error) {
-	unhashedSubpacketsLength := int(buf[0])<<8 | int(buf[1])
+	var unhashedSubpacketsLength int
+	if sig.Version == 6 {
+		unhashedSubpacketsLength = int(buf[0])<<24 | int(buf[1])<<16 | int(buf[2])<<8 | int(buf[3])
```
708c2ad5fb0cae90ddc282fd3967b30a6e9c72d9

wussler replied:
Well, yes :sweat_smile:

lubux commented on openpgp/packet/signature.go:
```diff
@@ -744,7 +948,30 @@ func (sig *Signature) SignUserId(id string, pub *PublicKey, priv *PrivateKey, co
+	return sig.Sign(h, priv, config)
+}
+
+// SignUserId computes a signature from priv
```
708c2ad5fb0cae90ddc282fd3967b30a6e9c72d9

lubux commented on openpgp/packet/signature.go:
```diff
@@ -830,6 +1073,12 @@ func (sig *Signature) Serialize(w io.Writer) (err error) {
+	if sig.Version == 6 {
+		// unhashed length is four-octet instead
+		// salt len 1 ocet
+		// len(salt) ocets
```
708c2ad5fb0cae90ddc282fd3967b30a6e9c72d9

lubux commented on openpgp/packet/signature.go:
```diff
@@ -842,18 +1091,41 @@ func (sig *Signature) Serialize(w io.Writer) (err error) {
-	fields := sig.HashSuffix[:6+hashedSubpacketsLen]
+	var fields []byte
+	if sig.Version == 6 {
+		// v6 signatures use 4 ocets for length
```
708c2ad5fb0cae90ddc282fd3967b30a6e9c72d9

lubux commented on openpgp/packet/signature.go:
```diff
@@ -1082,3 +1381,42 @@ func (sig *Signature) AddMetadataToHashSuffix() {
+// See https://datatracker.ietf.org/doc/html/draft-ietf-openpgp-crypto-refresh#section-9.5|Crypto Refresh Section 9.5
+func SaltLengthForHash(hash crypto.Hash) (int, error) {
+	switch hash {
+	case crypto.SHA256:
```
708c2ad5fb0cae90ddc282fd3967b30a6e9c72d9

lubux commented on openpgp/v2/canonical_text_test.go:
```diff
@@ -0,0 +1,52 @@
+// Copyright 2011 The Go Authors. All rights reserved.
```
It is a copy of the v1 code.

lubux commented on openpgp/v2/canonical_text.go:
```diff
@@ -0,0 +1,111 @@
+			if c == '\r' {
+				*s = 1
+			} else if c == '\n' {
+				cw.Write(buf[start:i])
```
Yes. Note that this code is just a copy of the existing v1 code but I fixed it in both places.
fb84e1605f230b6ffd21ebb863e1fa71730ec04c

wussler commented on openpgp/packet/config.go:
```diff
@@ -110,6 +131,21 @@ type Config struct {
+	CacheSessionKey bool
+	// CheckPacketSequence is a flag that controls if the pgp message reader should strictly check
+	// that the packet sequence conforms with the grammar mandated by rfc4880.
+	// The default behavior, when the config or flag is nil, is to check he packet sequence.
```
```suggestion
	// The default behavior, when the config or flag is nil, is to check the packet sequence.
```

lubux replied:
https://github.com/ProtonMail/go-crypto/pull/182/commits/26e7edcf30731cf2764f269e07445690253f2cc6

lubux commented on openpgp/v2/keys.go:
```diff
@@ -0,0 +1,787 @@
+		}
+	}
+	if len(primaryIdentityCandidates) == 0 {
+		return nil, nil, errors.StructuralError("no primary identity found")
```
Would actually makes sense to return nil and return an error in `PrimarySelfSignature` if a v4 keys has no identity.
Changed in: 78f1529721ab40073962b9b826edb0398c4686fb

lubux commented on openpgp/v2/keys.go:
```diff
@@ -0,0 +1,787 @@
+	return
+}
+
+// primarySelfSignature searches the entitity for the self-signature that stores key prefrences.
```
78f1529721ab40073962b9b826edb0398c4686fb

lubux commented on openpgp/v2/keys.go:
```diff
@@ -0,0 +1,787 @@
+	return k.PrimarySelfSignature == k.SelfSignature
+}
+
+// checkKeyRequirements
```
https://github.com/ProtonMail/go-crypto/commit/78f1529721ab40073962b9b826edb0398c4686fb

lubux commented on openpgp/v2/keys.go:
```diff
@@ -0,0 +1,787 @@
+		}
+		if usedKey.Version == 6 &&
+			(curve == packet.Curve25519 ||
+				curve == packet.Curve448) {
```
https://github.com/ProtonMail/go-crypto/commit/78f1529721ab40073962b9b826edb0398c4686fb

lubux commented on openpgp/v2/keys_v5_test.go:
```diff
@@ -0,0 +1,206 @@
+	}
+}
+
+// Depricated
```
https://github.com/ProtonMail/go-crypto/commit/78f1529721ab40073962b9b826edb0398c4686fb

lubux commented on openpgp/v2/read.go:
```diff
@@ -0,0 +1,791 @@
+				return 0, errors.StructuralError("parsing error")
+			}
+			if sig, ok := p.(*packet.Signature); ok {
+				if sig.Version == 5 && scr.md.LiteralData != nil && (sig.SigType == 0x00 || sig.SigType == 0x01) {
```
Yes, AFAIK file metadata is only include in the signature for v5.
v6 just ignores it.

lubux commented on openpgp/v2/read.go:
```diff
@@ -0,0 +1,791 @@
+	return md, nil
+}
+
+// checkSignatureDetails returns an error if:
```
78f1529721ab40073962b9b826edb0398c4686fb

lubux commented on openpgp/v2/key_generation.go:
78f1529721ab40073962b9b826edb0398c4686fb

lubux commented on openpgp/v2/subkeys.go:
```diff
@@ -0,0 +1,204 @@
+	return nil
+}
+
+func (s *Subkey) Serialize(w io.Writer, includeSecrets bool) error {
```
cc1afdc1d4fc3bc3ba1be4f8d06ff450f8d44224

lubux commented on openpgp/v2/write.go:
```diff
@@ -0,0 +1,1033 @@
+}
+
+// Sign signs a message. The resulting WriteCloser must be closed after the
+// contents of the file have been written.  hints contains optional information
```
3db51732b5b372aebda2697e4cdccbbbf3bb2fec

wussler replied:
There's still a space too much

lubux replied:
My bad: 8781dd323f5f58150fb65d04a5ff5cfb2ab1a9e7

lubux commented on openpgp/keys.go:
```diff
@@ -526,8 +526,35 @@ EachPacket:
-	if len(e.Identities) == 0 {
-		return nil, errors.StructuralError("entity without any identities")
+	if len(e.Identities) == 0 && e.PrimaryKey.Version < 6 {
+		return nil, errors.StructuralError("v4 entity without any identities")
```
dc7272a5ebae0567e24c41a766291d1f7a05c331

lubux commented on openpgp/keys.go:
```diff
@@ -526,8 +526,35 @@ EachPacket:
+		// Check that the main self-signature is valid
+		err = e.PrimaryKey.VerifyDirectKeySignature(mainDirectKeySelfSignature)
+		if err != nil {
+			return nil, errors.StructuralError("invalid direct-key self-signature for v6 xprimary key")
```
dc7272a5ebae0567e24c41a766291d1f7a05c331

lubux commented on openpgp/keys.go:
```diff
@@ -840,3 +879,23 @@ func (e *Entity) RevokeSubkey(sk *Subkey, reason packet.ReasonForRevocation, rea
+	return e.SelfSignature
+}
+
+// primarySelfSignature searches the entitity for the self-signature that stores key prefrences.
```
dc7272a5ebae0567e24c41a766291d1f7a05c331

lubux commented on openpgp/keys_test.go:
```diff
@@ -1514,7 +1514,6 @@ func TestEncryptAndDecryptPrivateKeys(t *testing.T) {
 		})
 	}
-	
 
```
dc7272a5ebae0567e24c41a766291d1f7a05c331

lubux commented:
@wussler thanks a lot for your time and effort in reviewing this slightly larger PR :sweat_smile:
I addressed all of your suggestions and added the commit as a comment in each thread. 

Just as a side note: The contents in the v2 folder are copies of the previous v1 root folder without the packages. 
So, the code is duplicated, and the new changes for v2 are integrated on top. 

I will do another pass as well.

wussler reviewed (approved):
Few remaining comments, then good to go for me :rocket:

lubux commented on openpgp/packet/private_key.go:
```diff
@@ -743,6 +964,86 @@ func (pk *PrivateKey) parseECDHPrivateKey(data []byte) (err error) {
+		err = errors.StructuralError("wrong X25519 key size")
+		return err
+	}
+	copy(privateKey.Secret, data)
```
Great catch, definitely a security issue.

lubux replied:
9cf78f4463f989d474f95545821e629e4a40b20e

lubux commented on openpgp/v2/read.go:
```diff
@@ -0,0 +1,795 @@
+
+func (sc *SignatureCandidate) validate() bool {
+	correspondingSig := sc.CorrespondingSig
+	invalidV3 := sc.OPSVersion == 3 && correspondingSig.Version == 6
```
I think its correct, see [link](https://www.ietf.org/archive/id/draft-ietf-openpgp-crypto-refresh-12.html#section-10.3.2.2). A v3 OPS packet is allowed in combination with a v3 or v4 signature packet.

wussler replied:
Bloody hell the v3 signatures, not PKESK - you're right

lubux commented on openpgp/v2/write.go:
```diff
@@ -0,0 +1,1053 @@
+		}
+	case packet.PubKeyAlgoECDSA:
+		if curve, err := singingKey.Curve(); err == nil {
+			if curve == packet.Curve448 ||
```
Uh indeed, fixed with: f2990fa8ed1a3dbf978e0e40228e7900a1bfea8f

lubux commented on openpgp/v2/read.go:
```diff
@@ -0,0 +1,795 @@
+import (
+	"bytes"
+	"crypto"
+	_ "crypto/sha256"
+	_ "crypto/sha512"
```
Should be, copied from [here](https://pkg.go.dev/crypto#Hash):
```
SHA256                      // import crypto/sha256
SHA384                      // import crypto/sha512
SHA512                      // import crypto/sha512
```

lubux commented on openpgp/v2/read.go:
```diff
@@ -0,0 +1,795 @@
+	HashAlgorithm     crypto.Hash
+	PubKeyAlgo        packet.PublicKeyAlgorithm
+	IssuerKeyId       uint64
+	IssuerFingerprint []byte // v6 only
```
ae457d10707840747bb0bea0c4bd81f35359cd17

lubux commented on openpgp/v2/read.go:
```diff
@@ -0,0 +1,795 @@
+// should be preprocessed (i.e. to normalize line endings). Thus this function
+// returns two hashes. The second should be used to hash the message itself and
+// performs any needed preprocessing.
+func hashForSignature(hashFunc crypto.Hash, sigType packet.SignatureType, sigSalt []byte) (hash.Hash, hash.Hash, error) {
```
ae457d10707840747bb0bea0c4bd81f35359cd17

lubux commented on openpgp/v2/keys_v5_test.go:
```diff
@@ -0,0 +1,202 @@
+	}
+}
+
+// Deprecated
```
0a030a354616ce1bec73a25ce8e2088acd9727ed

lubux commented on openpgp/packet/packet.go:
```diff
@@ -367,8 +370,93 @@ func Read(r io.Reader) (p Packet, err error) {
+	case packetTypeCompressed:
+		msgErr = sequence.Next(CompSymbol)
+		p = new(Compressed)
+	case packetTypeSymmetricallyEncrypted:
```
The `ReadWithCheck` function only checks if the packet sequence is correct. I think if we want to reject packets, we should do it at a higher level.

lubux commented on openpgp/packet/one_pass_signature.go:
```diff
@@ -7,34 +7,37 @@ package packet
+	KeyId          uint64
+	IsLast         bool
+	Salt           []byte // v6 only
+	KeyFingerprint []byte // v6 only
```
I think here it is correct because only v6 OPS packets contain a fingerprint of the signing key.

[section-5.4-3.7.1](https://www.ietf.org/archive/id/draft-ietf-openpgp-crypto-refresh-12.html#section-5.4-3.7.1)

lubux commented on openpgp/packet/encrypted_key.go:
```diff
@@ -284,3 +506,72 @@ func serializeEncryptedKeyECDH(w io.Writer, rand io.Reader, header [10]byte, pub
+func serializeEncryptedKeyX448(w io.Writer, rand io.Reader, header []byte, pub *x448.PublicKey, keyBlock []byte, cipherFunc byte, version int) error {
+	ephemeralPublicX448, ciphertext, err := x448.Encrypt(rand, pub, keyBlock)
+	if err != nil {
+		return errors.InvalidArgumentError("x448 encryption failed: " + err.Error())
```
Yes sure: 0b527c66ab238c771f59ea93a167acc93217e374

lubux commented on openpgp/keys_v5_test.go:
```diff
@@ -24,6 +20,8 @@ func TestReadPrivateForeignV5Key(t *testing.T) {
 	}
 }
 
+// Deprecated
```
https://github.com/ProtonMail/go-crypto/commit/0a030a354616ce1bec73a25ce8e2088acd9727ed
